Source/DafnyStandardLibraries/src/Std/DateTime/DateTimeUtils.dfy

```diff
@@ -0,0 +1,137 @@
+include "DateTimeConstant.dfy"
+
+module Std.DateTimeUtils {
+  import opened Strings
+  import opened BoundedInts
+  import opened DateTimeConstant
+
+  // Month names for better error messages
+  const MONTH_NAMES: seq<string> := ["January", "February", "March", "April", "May", "June",
+                                     "July", "August", "September", "October", "November", "December"]
+
+  // Function versions for use in function contexts
+  function {:extern "DateTimeImpl", "ToEpochTimeMilliseconds"}
+    {:axiom} ToEpochTimeMillisecondsFunc(year: int32, month: uint8, day: uint8, hour: uint8, minute: uint8, second: uint8, millisecond: uint16): int
+
+  function {:extern "DateTimeImpl", "FromEpochTimeMilliseconds"}
+    {:axiom} FromEpochTimeMillisecondsFunc(epochMillis: int): seq<int32>
+    ensures |FromEpochTimeMillisecondsFunc(epochMillis)| == 7
+    ensures var components := FromEpochTimeMillisecondsFunc(epochMillis);
+            IsValidDateTime(components[0], components[1] as uint8, components[2] as uint8, components[3] as uint8, components[4] as uint8, components[5] as uint8, components[6] as uint16)
+
+  // External method for getting current time components
+  method {:extern "DateTimeImpl", "GetNowComponents"}
+    {:axiom} GetNowComponents() returns (components: seq<int32>)
+    ensures |components| == 7
+
+  // Leap year calculation using extern implementation
+  function {:extern "DateTimeImpl", "IsLeapYear"}
+    {:axiom} IsLeapYear(year: int32): bool
+
+  // Days in month calculation
+  function DaysInMonth(year: int32, month: uint8): uint8
+    requires 1 <= month <= MONTHS_PER_YEAR
+  {
+    if month == 2 then
+      if IsLeapYear(year) then 29 else 28
+    else if month == 4 || month == 6 || month == 9 || month == 11 then 30
+    else 31
+  }
+
+  // Days in year calculation
+  function DaysInYear(year: int32): int
+  {
+    if IsLeapYear(year) then 366 else 365
+  }
+
+  // Month name getter
+  function GetMonthName(month: uint8): string
+    requires 1 <= month <= MONTHS_PER_YEAR
+  {
+    MONTH_NAMES[month - 1]
+  }
+
+  // Date-time validation predicate
+  predicate IsValidDateTime(year: int32, month: uint8, day: uint8, hour: uint8, minute: uint8, second: uint8, millisecond: uint16)
+  {
+    MIN_YEAR <= year <= MAX_YEAR &&
+    1 <= month <= MONTHS_PER_YEAR &&
+    1 <= day <= MAX_DAYS_PER_MONTH &&
+    0 <= hour < HOURS_PER_DAY &&
+    0 <= minute < MINUTES_PER_HOUR &&
+    0 <= second < SECONDS_PER_MINUTE &&
+    0 <= millisecond < MILLISECONDS_PER_SECOND
+  }
+
+
+  // Day of week calculation using Sakamoto's algorithm
+  function GetDayOfWeek(year: int32, month: uint8, day: uint8): int32
```

Comment:
This is a clever algorithm, but it's implicitly dependent on leap year calculation without using the extern, so let's leave it out.